File: java/value/trunk/src/test/java/org/ppwcode/value_III/time/interval/AbstractBeginEndTimeIntervalTest.java
/*<license>
  Copyright 2008, PeopleWare n.v.
  NO RIGHTS ARE GRANTED FOR THE USE OF THIS SOFTWARE, EXCEPT, IN WRITING,
  TO SELECTED PARTIES.
</license>*/

package org.ppwcode.value_III.time.interval;


import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.ppwcode.value_III.time.DateHelpers.le;
import static org.ppwcode.value_III.time.Duration.delta;

import java.util.ArrayList;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.ppwcode.value_III.time.Duration;


public class AbstractBeginEndTimeIntervalTest extends AbstractTimeIntervalTest {

  private static final class StubAbstractBeginEndTimeInterval extends AbstractBeginEndTimeInterval {

    public StubAbstractBeginEndTimeInterval(Date begin, Date end) throws IllegalIntervalException {
      super(begin, end);
    }

    public TimeInterval determinate(Date stubBegin, Date stubEnd) {
      return null;
    }

  }

  @SuppressWarnings("unchecked")
  @Override
  public List<? extends AbstractBeginEndTimeInterval> subjects() {
    return (List<? extends AbstractBeginEndTimeInterval>)$subjects;
  }

  @Override
  @Before
  public void setUp() throws IllegalIntervalException {
    List<AbstractBeginEndTimeInterval> s = new ArrayList<AbstractBeginEndTimeInterval>();
    final Date now = new Date();
    AbstractBeginEndTimeInterval subject = new StubAbstractBeginEndTimeInterval(now, null);
    s.add(subject);
    subject = new StubAbstractBeginEndTimeInterval(null, now);
    s.add(subject);
    subject = new StubAbstractBeginEndTimeInterval(now, now);
    s.add(subject);
    final GregorianCalendar gcB = new GregorianCalendar(2000, 0, 1);
    final GregorianCalendar gcE = new GregorianCalendar(2122, 11, 31);
    final Date b = gcB.getTime();
    final Date e = gcE.getTime();
    subject = new StubAbstractBeginEndTimeInterval(b, e);
    s.add(subject);
    $subjects = s;
  }

  protected void assertInvariants(AbstractBeginEndTimeInterval subject) {
    super.assertInvariants(subject);
  }

  @Test
  public void testGetDuration() {
    for (AbstractBeginEndTimeInterval subject : subjects()) {
      Duration result = subject.getDuration();
      Duration expected = subject.getBegin() == null || subject.getEnd() == null ? null :
        delta(subject.getBegin(), subject.getEnd());
      assertEquals(expected, result);
      assertInvariants(subject);
    }
  }

  @Test
  public void testAbstractBeginEndTimeInterval() {
    GregorianCalendar past = new GregorianCalendar(1995, 3, 24);
    Date now = new Date();
    GregorianCalendar future = new GregorianCalendar(2223, 4, 13);
    Date[] dates = new Date[] {null, past.getTime(), now, future.getTime()};
    for (Date d1 : dates) {
      for (Date d2 : dates) {
        try {
          AbstractBeginEndTimeInterval subject = new StubAbstractBeginEndTimeInterval(d1, d2);
          assertEquals(d1, subject.getBegin());
          assertEquals(d2, subject.getEnd());
          assertInvariants(subject);
        }
        catch (IllegalIntervalException exc) {
          assertTrue(! le(d1, d2) || (d1 == null && d2 == null));
        }
      }
    }
  }

}

File: dotnet/Value/trunk/src/Test_I/Time/Interval/AbstractBeginEndTimeIntervalTest.cs
/*<license>
Copyright 2011 - $Date: 2008-11-06 15:27:53 +0100 (Thu, 06 Nov 2008) $ by PeopleWare n.v..

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
</license>*/

#region Using

using System;
using System.Collections.Generic;
using System.Linq;

using Microsoft.VisualStudio.TestTools.UnitTesting;

using PPWCode.Value.I.Time.Interval;

#endregion

namespace PPWCode.Value.Test_I.Time.Interval
{
    [TestClass]
    public class AbstractBeginEndTimeIntervalTest : AbstractTimeIntervalTest
    {
        private class StubAbstractBeginEndTimeInterval : AbstractBeginEndTimeInterval
        {
            public StubAbstractBeginEndTimeInterval(DateTime? begin, DateTime? end)
                : base(begin, end)
            {
                // NOP
            }

            public override ITimeInterval Determinate(DateTime? stubBegin, DateTime? stubEnd)
            {
                return null;
            }
        }

        #region Additional test attributes

        //
        // You can use the following additional attributes as you write your tests:
        //
        // Use ClassInitialize to run code before running the first test in the class
        // [ClassInitialize()]
        // public static void MyClassInitialize(TestContext testContext) { }
        //
        // Use ClassCleanup to run code after all tests in a class have run
        // [ClassCleanup()]
        // public static void MyClassCleanup() { }
        //

        public IList<AbstractBeginEndTimeInterval> BeginEndSubjects
        {
            get
            {
                return Subjects.Cast<AbstractBeginEndTimeInterval>().ToList();
            }
        }

        [TestInitialize()]
        public override void MyTestInitialize()
        {
            List<AbstractTimeInterval> s = new List<AbstractTimeInterval>();
            DateTime now = DateTime.Now;
            AbstractBeginEndTimeInterval subject = new StubAbstractBeginEndTimeInterval(now, null);
            s.Add(subject);
            subject = new StubAbstractBeginEndTimeInterval(null, now);
            s.Add(subject);
            subject = new StubAbstractBeginEndTimeInterval(now, now);
            s.Add(subject);
            DateTime b = new DateTime(2000, 1, 1);
            DateTime e = new DateTime(2122, 12, 31);
            subject = new StubAbstractBeginEndTimeInterval(b, e);
            s.Add(subject);
            MSubjects = s;
        }

        #endregion

        [TestMethod]
        public void TestDuration()
        {
            TimeSpan? result = null;
            foreach (AbstractBeginEndTimeInterval subject in BeginEndSubjects)
            {
                result = subject.Duration;
            }
        }

        [TestMethod]
        public void TestConstructor()
        {
            DateTime past = new DateTime(1995, 4, 24);
            DateTime future = new DateTime(2223, 5, 13);
            DateTime now = DateTime.Now;
            DateTime?[] dates = {
                null, past, now, future
            };
            foreach (DateTime? d1 in dates)
            {
                foreach (DateTime? d2 in dates)
                {
                    try
                    {
                        AbstractBeginEndTimeInterval subject = new StubAbstractBeginEndTimeInterval(d1, d2);
                    }
                    catch (IllegalTimeIntervalException)
                    {
                        // NOP
                    }
                }
            }
        }
    }
}